File: custom_erpnext/custom_erpnext/report/monthly_leave_report/monthly_leave_report.js
// Copyright (c) 2022, Lithe-Tech Limited and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Monthly Leave Report"] = {
	"filters": [
		{
			"fieldname": "from_month",
			"label": __("From Month"),
			"fieldtype": "Select",
			"reqd": 1 ,
			"options": [
				{ "value": 1, "label": __("Jan") },
				{ "value": 2, "label": __("Feb") },
				{ "value": 3, "label": __("Mar") },
				{ "value": 4, "label": __("Apr") },
				{ "value": 5, "label": __("May") },
				{ "value": 6, "label": __("June") },
				{ "value": 7, "label": __("July") },
				{ "value": 8, "label": __("Aug") },
				{ "value": 9, "label": __("Sep") },
				{ "value": 10, "label": __("Oct") },
				{ "value": 11, "label": __("Nov") },
				{ "value": 12, "label": __("Dec") },
			],
			"default": frappe.datetime.str_to_obj(frappe.datetime.get_today()).getMonth() + 1
		},
		{
			"fieldname": "to_month",
			"label": __("To Month"),
			"fieldtype": "Select",
			"reqd": 1 ,
			"options": [
				{ "value": 1, "label": __("Jan") },
				{ "value": 2, "label": __("Feb") },
				{ "value": 3, "label": __("Mar") },
				{ "value": 4, "label": __("Apr") },
				{ "value": 5, "label": __("May") },
				{ "value": 6, "label": __("June") },
				{ "value": 7, "label": __("July") },
				{ "value": 8, "label": __("Aug") },
				{ "value": 9, "label": __("Sep") },
				{ "value": 10, "label": __("Oct") },
				{ "value": 11, "label": __("Nov") },
				{ "value": 12, "label": __("Dec") },
			],
			"default": frappe.datetime.str_to_obj(frappe.datetime.get_today()).getMonth() + 1
		},
		{
			"fieldname":"year",
			"label": __("Year"),
			// "default": frappe.datetime.get_year(),
			"fieldtype": "Select",
			"reqd": 1
		},
        {
            "fieldname": "company",
            "fieldtype": "Link",
            "label": "Company",
            "mandatory": 0,
            "options": "Company",
            "wildcard_filter": 0
        },

        {
            "fieldname": "employee",
            "fieldtype": "Link",
            "label": "Employee",
            "mandatory": 0,
            "options": "Employee",
            "wildcard_filter": 0
        },


        {
         "fieldname": "department",
         "fieldtype": "Link",
         "label": "Department",
         "mandatory": 0,
         "options": "Department",
         "wildcard_filter": 0
        },
        {
         "fieldname": "designation",
         "fieldtype": "Link",
         "label": "Designation",
         "mandatory": 0,
         "options": "Designation",
         "wildcard_filter": 0
        },
        {
            "fieldname": "section",
            "fieldtype": "Link",
            "label": "Section",
            "mandatory": 0,
            "options": "Section",
            "wildcard_filter": 0
           },	
           {
            "fieldname": "group_name",
            "fieldtype": "Link",
            "label": "Employee Group",
            "mandatory": 0,
            "options": "Group",
            "wildcard_filter": 0
           },
           
           {
            "fieldname": "facility_or_line",
            "fieldtype": "Link",
            "label": "Facility/Line",
            "mandatory": 0,
            "options": "Facility or Line",
            "wildcard_filter": 0

           },

           {
            "fieldname": "floor",
            "fieldtype": "Link",
            "label": "Floor",
            "mandatory": 0,
            "options": "Floor",
            "wildcard_filter": 0

           },
           

        {
            "fieldname": "shift",
            "fieldtype": "Link",
            "label": "Shift",
            "mandatory": 0,
            "options": "Shift Type",
            "wildcard_filter": 0
        }
	],
    "onload": function() {
		return  frappe.call({
			method: "erpnext.hr.report.monthly_attendance_sheet.monthly_attendance_sheet.get_attendance_years",
			callback: function(r) {
				var year_filter = frappe.query_report.get_filter('year');
				year_filter.df.options = r.message;
				year_filter.df.default = r.message.split("\n")[0];
				year_filter.refresh();
				year_filter.set_input(year_filter.df.default);
			}
		});
	}
};

File: custom_erpnext/custom_erpnext/report/monthly_leave_report/monthly_leave_report.py
# Copyright (c) 2022, Lithe-Tech Limited and contributors
# For license information, please see license.txt

# import frappe

from calendar import monthrange
from datetime import datetime
import frappe
from frappe import _, msgprint
from frappe.utils import cint, cstr, getdate

def execute(filters=None):
	
	if not filters:
		filters = {}
	columns = get_columns()
	data = get_missing_entry(filters)
	
	return columns, data
def get_columns():
	return [
		_("Emp Id") + ":Link/Employee:120",
		_("Designation") + ":Data/:120",
		# _("Section") + ":Data/:120",
		_("Approved date") + ":Data/:120",
		_("From Date") + ":Data/:120",
		_("To Date") + ":Data/:120",
		_("Leave Type") + ":Data/:120",
		_("Leave Days") + ":Data/:120",
		_("Remarks") + ":Data/:120"
    ]

def get_missing_entry(filters):
	conditions, filters = get_conditions(filters)
	result= frappe.db.sql("""select la.employee, emp.designation, la.posting_date, la.from_date, la.to_date, la.leave_type,
		la.total_leave_days, la.description
		FROM `tabLeave Application` la
		LEFT JOIN tabEmployee emp ON la.employee = emp.name  
		where 
		la.docstatus<2 %s
		""" 
		%conditions, 
		as_list=1)
	return result

def get_conditions(filters):
	conditions="" 
	if not (filters.get("to_month") and filters.get("year")):
		msgprint(_("Please select month and year"), raise_exception=1)

	total_days_in_to_month = monthrange(cint(filters.year), cint(filters.to_month))[1]
	to_date=filters.get("year")+"-"+str(filters.get("to_month")).zfill(2)+"-"+str(total_days_in_to_month)
     
	from_date=filters.get("year")+"-"+str(filters.get("from_month")).zfill(2)+"-"+"01"
	#from_date1=datetime.strptime(from_date,'%y-%m-%d')

	# if filters.get("from_date"): conditions += " and att.attendance_date>= '%s'" % filters["from_date"]
	# if filters.get("to_date"): conditions += " and att.attendance_date<= '%s'" % filters["to_date"]

	if filters.get("company"): conditions += " and la.company = '%s'" % filters["company"]	
	#if filters.get("employee"): conditions += " and att.employee= '%s'" % filters["employee"]
	if filters.get("department"): conditions += " and la.department= '%s'" % filters["department"]
	if filters.get("designation"): conditions += " and emp.designation='%s'" % filters["designation"]
	# if filters.get("section"): conditions += " and emp.section='%s'" % filters["section"]
	# if filters.get("floor"): conditions += " and emp.floor='%s'" % filters["floor"]
	# if filters.get("facility_or_line"): conditions += " and emp.facility_or_line='%s'" % filters["facility_or_line"]
	# if filters.get("group_name"): conditions += " and emp.group='%s'" % filters["group_name"]
	conditions += " and la.from_date > '%s' and la.to_date< '%s'" %(from_date,to_date)    ##It works


	return conditions, filters
